ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -622,3 +630,126 @@ void trackAgeAtAccess(long creationTimeMs) {
+    requestProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + REQUEST_PROCESSING_TIME_SUFFIX));
+    responseProcessingWaitTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_WAIT_TIME_SUFFIX));
+    responseProcessingTimeInMs =
+        metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + RESPONSE_PROCESSING_TIME_SUFFIX));
+    roundTripTimeInMs = metricRegistry.histogram(MetricRegistry.name(ownerClass, requestType + ROUND_TRIP_TIME_SUFFIX));
+
+    operationRate = metricRegistry.meter(MetricRegistry.name(ownerClass, requestType + OPERATION_RATE_SUFFIX));
+    operationErrorRate = metricRegistry.meter(
+        MetricRegistry.name(ownerClass, requestType + OPERATION_ERROR_SUFFIX + OPERATION_RATE_SUFFIX));
+  }
+}
+
+/**
+ * Construct to support end-to-end metrics tracking for crypto jobs. Usually accompanies a single
+ * {@link CryptoJob} i.e. there is a one-to-one mapping b/w a {@link CryptoJob} and a CryptoJobMetricsTracker
+ * instance.
+ * <p/>
+ * A brief description of how the tracker works :-
+ * - When an object of type {@link CryptoJob} is instantiated, it is also expected to be associated with a
+ *    unique instance of CryptoJobMetricsTracker (tracker). The Tracker will be instantiated either with
+ *    EncryptJobMetrics or DecryptJobMetrics
+ * - As the cryptoJob passed through various phases, metrics associated with these are updated via methods exposed
+ *    by this tracker.
+ */
+class CryptoJobMetricsTracker {
+  private final CryptoJobMetrics cryptoJobMetrics;
+  private long trackerStartTimeMs;
+  private long lastSeenTimeMs;
+
+  CryptoJobMetricsTracker(CryptoJobMetrics cryptoJobMetrics) {
+    this.cryptoJobMetrics = cryptoJobMetrics;
+  }
+
+  /**
+   * Starts the operation tracker for the corresponding CryptoJob
+   */
+  void startTracker() {
+    trackerStartTimeMs = System.currentTimeMillis();
+    lastSeenTimeMs = trackerStartTimeMs;
+  }
+
+  /**
+   * Notifies that the request processing has started for the corresponding CryptoJob
+   */
+  void startRequestProcessing() {
+    cryptoJobMetrics.operationRate.mark();
+    lastSeenTimeMs = System.currentTimeMillis();
+    cryptoJobMetrics.requestQueuingTimeInMs.update(lastSeenTimeMs - trackerStartTimeMs);
+  }
+
+  /**
+   * Notifies that the request processing has completed for the corresponding CryptoJob
+   */
+  void completeRequestProcessing() {
+    long curTimeMs = System.currentTimeMillis();
+    cryptoJobMetrics.requestProcessingTimeInMs.update(curTimeMs - lastSeenTimeMs);
+    lastSeenTimeMs = curTimeMs;
+  }
+
+  /**
+   * Notifies that the response processing has started for the corresponding CryptoJob
+   */
+  void startResponseProcessing() {
+    long curTimeMs = System.currentTimeMillis();
+    cryptoJobMetrics.responseProcessingWaitTimeInMs.update(curTimeMs - lastSeenTimeMs);
+    lastSeenTimeMs = curTimeMs;
```

Comment:
IMO, using this "lastSeenTime" is very confusing. Why not have "requestProcessingStartTime", "responseProcessingStartTime" etc? (also I don't think "request" and "response" is the correct terminology)